### PR TITLE
Fix incorrect functionality of ClientSpectatePacket

### PR DESCRIPTION
**Pull request by KodingDev**
See code for an explanation of why this is needed.

### CODE REVIEW

**Comment by Moulberry** [comment]
Firstly, I wouldn't implement this as a default feature, since not everyone using Minestom would want this behaviour.
I would recommend adding a "SpectatorTeleportEvent," then people can opt-in to this behavior by implementing it themselves.

Furthermore, when implementing default features into a library, you need to be exceedingly careful of not enabling exploits.
Your implementation allows a spectator to teleport to any player on the server, even if the spectating player cannot see them.
A hacker could use this to teleport to an admin (who might be in a special admin world)

**Review by Moulberry — CHANGES_REQUESTED** [review]
Firstly, I wouldn't implement this as a default feature, since not everyone using Minestom would want this behaviour.
I would recommend adding a "SpectatorTeleportEvent," then people can opt-in to this behavior by implementing it themselves.

Furthermore, when implementing default features into a library, you need to be exceedingly careful of not enabling exploits.
Your implementation allows a spectator to teleport to any player on the server, even if the spectating player cannot see them.
A hacker could use this to teleport to an admin (who might be in a special admin world)

File: src/main/java/net/minestom/server/listener/SpectateListener.java [hunk 1]
```diff
@@ -13,11 +13,22 @@ public static void listener(ClientSpectatePacket packet, Player player) {
...
             return;
+        }
 
-        // TODO check if 'target' is in a different instance
-        player.spectate(target);
+        if (target.getInstance() == null || player.getInstance() == null) {
+            return;
+        }
+
+        if (target.getInstance().getUniqueId() != player.getInstance().getUniqueId()) {
```
Comment (Moulberry) [inline src/main/java/net/minestom/server/listener/SpectateListener.java:24]: Won't this always return?

Comment (KodingDev) [reply src/main/java/net/minestom/server/listener/SpectateListener.java:24]: If they're in the same instance they'll bypass this check, i see no reason for it to always return.

**Comment by KodingDev** [comment]
> Firstly, I wouldn't implement this as a default feature, since not everyone using Minestom would want this behaviour. I would recommend adding a "SpectatorTeleportEvent," then people can opt-in to this behavior by implementing it themselves.
> 
> Furthermore, when implementing default features into a library, you need to be exceedingly careful of not enabling exploits. Your implementation allows a spectator to teleport to any player on the server, even if the spectating player cannot see them. A hacker could use this to teleport to an admin (who might be in a special admin world)

Commenting on this, this is just fixing a bug over anything else. The same issues mentioned apply to the previous implementation. The issue is that the spectate function is called, but there is no default logic allowing a player to exit from spectating, so they're stuck in that state until they relog. The main purpose of this PR is to correct this to the Vanilla implementation which is just to teleport the spectator to the player.

Adjusting the implementation of this brings along the question of if you want to allow the other default listeners to be disabled too, which seems out of scope for a bugfix PR like this.

**Review by Moulberry — APPROVED** [review]
lgtm

File: src/main/java/net/minestom/server/listener/SpectateListener.java [hunk 2]
```diff
@@ -1,23 +1,45 @@
...
         final Entity target = Entity.getEntity(targetUuid);
 
         // Check if the target is valid
-        if (target == null || target == player)
+        if (target == null || target == player) {
+            return;
+        }
+
+        // Ignore if they're not attached to any instances
+        if (target.getInstance() == null || player.getInstance() == null) {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/listener/SpectateListener.java:29]: Store the instances as variable and reuse them instead.

File: src/main/java/net/minestom/server/network/packet/client/play/ClientSpectatePacket.java [hunk 3]
```diff
@@ -7,6 +7,15 @@
...
 
+/**
+ * The ClientSpectatePacket is sent when the client interacts with their hotbar to switch between entities.
+ * Contrary to its name, it is actually used to teleport the player to the entity they are switching to,
+ * rather than spectating them.
+ * <p>
+ * Minestom will, by default, handle this behavior for you.
+ *
+ * @param target The UUID of the entity the player is switching to.
+ */
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/client/play/ClientSpectatePacket.java:18]: The doc seems unnecessary here, it is also not something we want to maintain

**Review by TheMode — CHANGES_REQUESTED** [review]

**Pull request merged**